packages/edit-navigation/src/components/menu-editor/use-create-missing-menu-items.js

```diff
@@ -17,41 +18,95 @@ import PromiseQueue from './promise-queue';
-			if ( enqueuedBlocksIds.current.includes( clientId ) ) {
-				continue;
-			}
-			enqueuedBlocksIds.current.push( clientId );
-			promiseQueueRef.current.enqueue( () =>
-				createDraftMenuItem( clientId ).then( ( menuItem ) => {
-					menuItemsRef.current[ clientId ] = menuItem;
-					enqueuedBlocksIds.current.splice(
-						enqueuedBlocksIds.current.indexOf( clientId )
-					);
-				} )
-			);
+export default function useCreateMissingMenuItems( query ) {
+	const storeSavedMenuItem = useStoreSavedMenuItem( query );
+	return useProcessNewBlocks(
+		query,
+		createDraftMenuItem,
+		storeSavedMenuItem
+	);
+}
+
+function useProcessNewBlocks( query, processClientId, onClientIdProcessed ) {
+	const promiseQueueRef = useRef();
+	const processedClientIdsRef = useRef();
+	useEffect( () => {
+		if ( promiseQueueRef.current ) {
+			promiseQueueRef.current.halt();
 		}
-	};
-	const onCreated = useCallback(
+		promiseQueueRef.current = new PromiseQueue( 5 );
+		processedClientIdsRef.current = [];
+	}, [ query ] );
+
+	const processNewBlocks = useCallback(
+		( previousBlocks, newBlocks ) => {
+			const promiseQueue = promiseQueueRef.current;
+			const processedClientIds = processedClientIdsRef.current;
+			const delta = diffBlocks( previousBlocks, newBlocks );
+			for ( const { clientId } of delta ) {
+				if ( processedClientIds.includes( clientId ) ) {
+					continue;
+				}
+				processedClientIds.push( clientId );
+				promiseQueue.enqueue( () =>
+					processClientId( clientId ).then( ( result ) => {
+						if ( promiseQueue.halted ) {
+							return;
+						}
+						onClientIdProcessed( clientId, result );
+					} )
+				);
+			}
+		},
+		[ query, processClientId, onClientIdProcessed ]
+	);
+	const onProcessed = useCallback(
 		( callback ) => promiseQueueRef.current.then( callback ),
 		[ promiseQueueRef.current ]
 	);
-	return { createMissingMenuItems, onCreated };
+	return [ processNewBlocks, onProcessed ];
+}
+
+function* diffBlocks( previousBlocks, newBlocks ) {
+	const existingClientIds = flattenBlocks( previousBlocks ).map(
+		( { clientId } ) => clientId
+	);
+	for ( const block of flattenBlocks( newBlocks ) ) {
+		const { clientId, name } = block;
+		// No need to create menuItems for the wrapping navigation block
+		if ( name === 'core/navigation' ) {
+			continue;
+		}
+		// Menu item was already created
+		if ( existingClientIds.includes( clientId ) ) {
+			continue;
+		}
+		yield block;
+	}
+}
+
+function useStoreSavedMenuItem( query ) {
+	const { assignMenuItemIdToClientId } = useDispatch(
+		'core/edit-navigation'
+	);
+	const { receiveEntityRecords } = useDispatch( 'core' );
+	const select = useSelect( ( s ) => s );
+	return useCallback(
+		( clientId, menuItem ) => {
+			assignMenuItemIdToClientId( query, menuItem.id, clientId );
+			receiveEntityRecords(
+				'root',
+				'menuItem',
+				[ ...select( 'core' ).getMenuItems( query ), menuItem ],
+				query,
+				false
+			);
+		},
+		[ query ]
+	);
```

Comment:
I am not too happy about this part. Previously `menuItems` were kept in react ref. Now the core data layer is the primary source of truth. This means that creating a temporary menu item should be reflected in redux state. The only way to do it ATM is `receiveEntityRecords` which accepts an entire page of results. Maybe we should add an action like `appendEntityRecord` to avoid this `select` madness here?